### PR TITLE
Use Request/Response pattern for Vectors layer

### PR DESCRIPTION
**Pull request by brisvag**
# Description
Part of the async work (#4795), bring `Vectors` up to speed. Mostly mirrors `Points`.

### CODE REVIEW

File: napari/layers/vectors/_slice.py [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+            return _VectorSliceResponse(
+                indices=np.arange(len(self.data), dtype=int),
+                alphas=1,
+                dims=self.dims,
+            )
+
+        # We want a numpy array so we can use fancy indexing with the non-displayed
+        # indices, but as self.dims_indices can (and often/always does) contain slice
+        # objects, the array has dtype=object which is then very slow for the
+        # arithmetic below. As Vectors._round_index is always False, we can safely
```
Comment (andy-sweet) [inline napari/layers/vectors/_slice.py:87]: Unlike `Points`, `Vectors._round_index` is `True`, so the comment is a bit misleading here. But maybe it should be `False` given that slicing acts in a point/vertex-like way.

In fact, I'd probably argue that it's just `_ImageBase` which should have `_round_index` be `True`. But I'll leave that discussion for later!

File: napari/layers/vectors/vectors.py [hunk 2]
```diff
@@ -593,83 +598,57 @@ def _view_face_color(self) -> np.ndarray:
...
-        return slice_indices, alpha
+        # TODO: [see Image]
+        #   For the existing sync slicing, slice_indices is passed through
+        # to avoid some performance issues related to the evaluation of the
+        # data-to-world transform and its inverse. Async slicing currently
+        # absorbs these performance issues here, but we can likely improve
+        # things either by caching the world-to-data transform on the layer
+        # or by lazily evaluating it in the slice task itself.
+        slice_indices = slice_input.data_indices(
+            self._data_to_world.inverse, round_index=False
```
Comment (andy-sweet) [inline napari/layers/vectors/vectors.py:669]: `round_index=False` is a change of behavior (though probably a wanted one). See my other comment: https://github.com/napari/napari/pull/5918/files#r1225443605 which is realize is wrong now :D

**Review by andy-sweet — COMMENTED** [review]
This generally looks good, though I still need to find some time to pull this down and try it out before approving.

A few of high level optional suggestions/comments.

1\. Consider adding a test to `test_async_slicing`. Most of the slicing code paths are shared between sync/async, but not entirely (yet), so this would be a good sanity check.

2\. I think `Points` can potentially break with async slicing right now when the data length gets smaller during slicing because slicing for points basically means finding indices. I think `Vectors` suffers from the same problem since the code is very similar. I don't we think we need to fix it here unless you feel strongly because while we will get an error, I think it will be self-corrective in most cases because changing data typically means re-slicing.

One other idea is to share more code between `Points` and `Vectors`, ideally through utility functions as described in https://github.com/napari/napari/issues/4896

**Review by andy-sweet — APPROVED** [review]
Tried this out and seems to work. The other comments are optional, but might be worth at least adding one test.

**Comment by brisvag** [comment]
> Unlike Points, Vectors._round_index is True, so the comment is a bit misleading here. But maybe it should be False given that slicing acts in a point/vertex-like way.

Yeah, as you noticed I also changed `Vectors` to use `round_index=False`, cause I assumed it was probably meant to be like that.

Will add t

> One other idea is to share more code between Points and Vectors, ideally through utility functions as described in https://github.com/napari/napari/issues/4896

Yeah that's probably the best course of actions; I wanted to see where the patterns were before doing a refactor. I think it would be especially useful to try this on a more different layer (like shapes) to see how much could be shared in general.

File: napari/layers/vectors/_slice.py [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+
+@dataclass(frozen=True)
+class _VectorSliceResponse:
+    """Contains all the output data of slicing an Vectors layer.
+
+    Attributes
+    ----------
+    indices : array like
+        Indices of the sliced Vectors data.
+    alphas: array like or scalar
```
Comment (Carreau) [inline napari/layers/vectors/_slice.py]:
```suggestion
    alphas : array like or scalar
```

**Comment by brisvag** [comment]
Added a test mimicking the points layer. A bit trickier cause we need to compare with the generated mesh, so I reused some of the implementation to maintain the intent behind the test.

**Comment by andy-sweet** [comment]
The test failure is interesting: https://github.com/napari/napari/actions/runs/5465012122/jobs/9947840076?pr=5918#step:8:195

Pretty certain it has nothing in particular to do with this PR (so shouldn't blocking merging it), but it does make me think I should disconnect `QtViewer._on_slice_ready` from `LayerSlicer.events.ready` in `Viewer.close` to break the cyclic dependency there. Or equivalently connect a weak version of `QtViewer._on_slice_ready`.

**Pull request merged**